### PR TITLE
Sphinx documentation in doc folder

### PR DESCRIPTION
**Pull request by engrosamaali91**
**Sphinix documentation **
- All the source files have been added to the doc folder as per LaPy documentation
- The mains files are configuration file i.e conf.py and index.rst which the main page of BrainPrint project

**What steps I followed to generate .rst files**
1. Git clone BrainPrint repository
2. Modified configuration file i.e **conf.py**, added relevant extension. (Some lines have been commented out such as sphinx_gallery_conf = {} as they were generating warnings and errors
3. Added relevant links to links.inc file
4. Generated rst files corresponding to python files and packages 
> cd docs
>  sphinix-apidoc -o (current directory) (source directory)

5. This sphinix-apidoc creates these rst files
> brainprint.cli.rst
> brainprint.commands.rst
> brainprint.rst
> brainprint.utils.rst
> brainprint.utils.tests.rst
> modules.rst

6. Then copied the brainprint.rst files details to the my toctree file i.e _api-> index.rst->bp.rst_
7. Built html files using _make html_

Note: The source files are not getting linked with corresponding github source code due to errors in step_2

### CODE REVIEW

**Comment by m-reuter** [comment]
Hi, I wonder why there is no numpydoc in the extensions? And why you put so many rst files manually instead of autogeneration? Compare to how it is done in LaPy (also templates are missing)?

**Comment by engrosamaali91** [comment]
Hello, I believe the sphinx-API doc in step-4 reads the .py files with docstring which I why I could see numpy docstring in the source code. Sphinx api doc command creates as many rst files as many python packages are there in the file. In brinprint we have utilis, commands, brainprint class, asymmetry, and onemore module. So sphinx api created all the corresponding rst files automatically, then its up to us which rst file we would like to add to our project. This is another method of autogeneration of rst files which I used instead of the one we used in LapY project. The reason why I used this method is that I was getting loads of warnings and errors while building html page with the LaPy method. I configured the rest as per LaPy method.

**Comment by engrosamaali91** [comment]
In commit 9170585876560ec15d80f8a5056fb8d7a1ba8883 the absolute path(local) from sphinx gallery configuration has been removed so that the document can be built on any version control systems like Git.

Please refer to line **no. 234** of conf.py file.

**Comment by engrosamaali91** [comment]
In commit [f0122ea](https://github.com/Deep-MI/BrainPrint/pull/13/commits/f0122ea742918636f1303127aed891134dd5c7c2) index.rst file has been modfied as per readme.me content.

Convert README.md file which was in markdown format to index.rst file which is in restructured text.

With direct parsing of readme.md file the sphinix is unable to recognize .md format and therefore it does not generate proper headings and cover entire long text in oneline. It also does not recognize clickable pypi image in the start.

**Comment by engrosamaali91** [comment]
- In commit [f957a5e](https://github.com/Deep-MI/BrainPrint/pull/13/commits/f957a5e9de230321229089b543767ffbf6fc3669) sphinix gallary has been commented out in conf.py file
- In commit [c12843f](https://github.com/Deep-MI/BrainPrint/pull/13/commits/c12843f4dc4c85cf961848888430cb68d9279e41) api referencing structure has been changed and placed under API reference only.

File: doc/references.bib [hunk 1]
```diff
@@ -0,0 +1,34 @@
+@article{conformal_parameterization_2020,
+ author = {Choi, Gary P. T. and Leung-Liu, Yusan and Gu, Xianfeng and Lui, Lok Ming},
+ doi = {10.1137/19M125337X},
+ journal = {SIAM Journal on Imaging Sciences},
+ number = {3},
+ pages = {1049-1083},
+ title = {Parallelizable Global Conformal Parameterization of Simply-Connected Surfaces via Partial Welding},
+ volume = {13},
+ year = {2020}
+}
```
Comment (m-reuter) [inline doc/references.bib:10]: This reference should be dropped as it is only relevant for Lapy and not Brainprint

**Comment by m-reuter** [comment]
I wonder if we can find out a way to directly include the README.md rather than having a copy in rst . Currently we need to remember to edit both files if we change anything, which is not great.

**Comment by m-reuter** [comment]
black fails on conf.py

File: doc/conf.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        filename = "/" + filename
+
+    if "fullname" in info:
+        anchor = info["fullname"]
+        anchor = "#:~:text=" + quote(anchor.split(".")[-1])
+    else:
+        anchor = ""
+
+    # Replace <gh_url> with the actual GitHub repository URL
+    gh_url = "https://github.com/Deep-MI/BrainPrint"  # Replace with your repository URL
```
Comment (m-reuter) [inline doc/conf.py:178]: gh_url is set as a variable in line 4, why is that not enough?

Comment (engrosamaali91) [reply doc/conf.py]: Indeed gh_url in line 24 is enough. Line 178 has been commented out. 46c220062f5020bd77415387859f7a449479c3cd

**Comment by engrosamaali91** [comment]
> I wonder if we can find out a way to directly include the README.md rather than having a copy in rst . Currently we need to remember to edit both files if we change anything, which is not great.

True, I actually tried including README.md file directly into index.rst file but .rst file did not properly render and import .md file content and it was not properly formatted as per readme.md file. I will look more into other methods to include README.md file.

**Comment by engrosamaali91** [comment]
> black fails on conf.py

conf.py file has been reformatted using black. 521ad6c38e62742bd7a2a81c54c11d3c05735fcb

**Comment by m-reuter** [comment]
Conf.py still fails in black. Maybe gh action uses a different version of  black?

**Comment by engrosamaali91** [comment]
> Conf.py still fails in black. Maybe gh action uses a different version of black?

Hello,
The linkcode_resolve function is now identical to the one used in LaPy. I have now commented out the linkcode_resolve function on line 142 because the linkcode_resolve used in LaPy wasn't properly linking the corresponding source code on GitHub after building for BrainPrint. However, it seems that the linking is now working as expected using LaPy linkcode_resolve function.

Regarding the issue with conf.py failing in black, I am using the same version (23.7.0) of black as GitHub. Additionally, I've ensured that the conf.py adheres to the configuration defined in 'Brainprint/pyproject.toml' by running black --verbose conf.py. My local and GitHub Actions configurations are in sync." d34a72390654fbcf87c9ff3d1e661858959fc68e

**Pull request merged**